drivers/wifi/infineon/airoc_whd_hal.c

```diff
@@ -0,0 +1,474 @@
+	return WHD_SUCCESS;
+}
+
+whd_result_t whd_bus_sdio_cmd53(whd_driver_t whd_driver, whd_bus_transfer_direction_t direction,
+				whd_bus_function_t function, sdio_transfer_mode_t mode,
+				uint32_t address, uint16_t data_size, uint8_t *data,
+				sdio_response_needed_t response_expected, uint32_t *response)
+{
+	whd_result_t result;
+	struct sd_card *sd = (struct sd_card *)whd_driver->bus_priv->sdio_obj;
+	sdio_cmd_argument_t arg = {0};
+
+	if (direction == BUS_WRITE) {
+		WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd53_write);
+	}
+
+	if (direction == BUS_READ) {
+		WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, cmd53_read);
+	}
+
+	arg.cmd53.function_number = (uint32_t)(function & BUS_FUNCTION_MASK);
+	arg.cmd53.register_address = (uint32_t)(address & BIT_MASK(17));
+	arg.cmd53.op_code = (uint32_t)1;
+	arg.cmd53.rw_flag = (uint32_t)((direction == BUS_WRITE) ? 1 : 0);
+
+	if (mode == SDIO_BYTE_MODE) {
+		__ASSERT(data_size <= (uint16_t)512,
+			 '"%s...": whd_bus_sdio_cmd53: data_size > 512 for byte mode', __func__);
+		arg.cmd53.count = (uint32_t)(data_size & 0x1FF);
+
+	} else {
+		arg.cmd53.count = (uint32_t)((data_size / (uint16_t)SDIO_64B_BLOCK) & BIT_MASK(9));
+		if ((uint32_t)(arg.cmd53.count * (uint16_t)SDIO_64B_BLOCK) < data_size) {
+			++arg.cmd53.count;
+		}
+		arg.cmd53.block_mode = 1;
+	}
+
+	struct sdhc_command sdhc_cmd = {
+		.opcode = SDIO_RW_EXTENDED,
+		.response_type = SD_RSP_TYPE_R5,
+		.arg = arg.value,
+	};
+
+	struct sdhc_data sdhc_data = {
+		.timeout_ms = CONFIG_SD_DATA_TIMEOUT,
+		.data = data,
+		.block_size = (mode == SDIO_BYTE_MODE) ? data_size : SDIO_64B_BLOCK,
+		.blocks = (mode == SDIO_BYTE_MODE) ? 1 : arg.cmd53.count,
+	};
+	result = sdhc_request(sd->sdhc, &sdhc_cmd, &sdhc_data);
+
+	WHD_BUS_STATS_CONDITIONAL_INCREMENT_VARIABLE(
+		whd_driver->bus_priv, ((result != WHD_SUCCESS) && (direction == BUS_READ)),
+		cmd53_read_fail);
+	WHD_BUS_STATS_CONDITIONAL_INCREMENT_VARIABLE(
+		whd_driver->bus_priv, ((result != WHD_SUCCESS) && (direction == BUS_WRITE)),
+		cmd53_write_fail);
+
+	CHECK_RETURN(result);
+	return WHD_SUCCESS;
+}
+
+/*
+ * Implement SDIO Card interrupt
+ */
+
+void whd_bus_sdio_irq_handler(const struct device *dev, int reason, const void *user_data)
+{
+	if (reason == SDHC_INT_SDIO) {
+		whd_driver_t whd_driver = (whd_driver_t)user_data;
+
+		WHD_BUS_STATS_INCREMENT_VARIABLE(whd_driver->bus_priv, sdio_intrs);
+
+		/* call thread notify to wake up WHD thread */
+		whd_thread_notify_irq(whd_driver);
+	}
+}
+
+whd_result_t whd_bus_sdio_irq_register(whd_driver_t whd_driver)
+{
+	/* Nothing to do here, all handles by whd_bus_sdio_irq_enable function */
+	return WHD_SUCCESS;
+}
+
+whd_result_t whd_bus_sdio_irq_enable(whd_driver_t whd_driver, whd_bool_t enable)
+{
+	struct sd_card *sd = (struct sd_card *)whd_driver->bus_priv->sdio_obj;
+
+	/* Enable/disable SDIO Card interrupts */
+	if (enable) {
+		sdhc_enable_interrupt(sd->sdhc, whd_bus_sdio_irq_handler, SDHC_INT_SDIO,
```

Comment:
sdhc_enable_interrupt returns a status code, so needs to be checked.

Reply:
updated